frame/dapps-staking/src/pallet/mod.rs

```diff
@@ -965,34 +968,45 @@ pub mod pallet {
             GeneralEraInfo::<T>::insert(era, era_info);
         }
 
-        /// This helper returns `ContractStakeInfo` for given era if possible or latest stored data
-        /// or finally default value if storage have no data for it.
+        /// Used to copy all `ContractStakeInfo` from the ending era over to the next era.
+        /// This is the most primitive solution since it scales with number of dApps.
+        /// It is possible to provide a hybrid solution which allows laziness but also prevents
+        /// a situation where we don't have access to the required data.
+        fn advance_staking_info(current_era: EraIndex) -> u64 {
+            let next_era = current_era + 1;
+
+            let mut consumed_weight = 0;
+
+            for (contract_id, dapp_info) in RegisteredDapps::<T>::iter() {
+                // Ignore dapp if it was unregistered
+                if let DAppState::Unregistered(_) = dapp_info.state {
+                    consumed_weight = consumed_weight.saturating_add(T::DbWeight::get().reads(1));
+                    continue;
+                }
+
+                // Copy data from era `X` to era `X + 1`
+                if let Some(mut staking_info) =
+                    ContractEraStake::<T>::get(&contract_id, current_era)
+                {
+                    staking_info.contract_reward_claimed = false;
+                    ContractEraStake::<T>::insert(&contract_id, next_era, staking_info);
+
+                    consumed_weight =
+                        consumed_weight.saturating_add(T::DbWeight::get().reads_writes(1, 1));
+                } else {
+                    consumed_weight = consumed_weight.saturating_add(T::DbWeight::get().reads(1));
+                }
+            }
+
+            consumed_weight
+        }
+
+        /// This helper returns `ContractStakeInfo` for given era or default value if storage have no data for it.
         pub fn staking_info(
```

Comment:
May be we can drop this helper because it become just an alias for storage read call.

Reply:
Sure, I kept it just in order to keep changes for this PR version to the minimum.